### ISSUE TITLE
Adding an object that contains a firestore.DocumentReference to an Akita store throws an exception

### ISSUE THREAD

**Issue by dylanwatsonsoftware** [datorama/akita#354]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Current behaviour

Adding a firebase object, that contains a DocumentReference, to an Akita store throws an exception:
<pre>
<code>
ERROR Error: Uncaught (in promise): TypeError: Cannot assign to read only property 'operationInProgress' of object '[object Object]'
TypeError: Cannot assign to read only property 'operationInProgress' of object '[object Object]'
    at index.cjs.js:1629
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (zone.js:391)
    at Object.onInvoke (core.js:24349)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (zone.js:390)
    at Zone.push../node_modules/zone.js/dist/zone.js.Zone.run (zone.js:150)
    at zone.js:910
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:423)
    at Object.onInvokeTask (core.js:24340)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:422)
    at Zone.push../node_modules/zone.js/dist/zone.js.Zone.runTask (zone.js:195)
    at drainMicroTaskQueue (zone.js:601)
    at push../node_modules/zone.js/dist/zone.js.ZoneTask.invokeTask (zone.js:502)
    at ZoneTask.invoke (zone.js:487)
    at timer (zone.js:3070)
    at resolvePromise (zone.js:852) [angular]
    at :4200/polyfills.js:5666:17 [angular]
    at Object.onInvokeTask (core.js:24340) [angular]
    at drainMicroTaskQueue (zone.js:601) [<root>]
    at push../node_modules/zone.js/dist/zone.js.ZoneTask.invokeTask (zone.js:502) [<root>]
    at timer (zone.js:3070) [<root>]
</code> 
</pre>

## Expected behavior

Ideally, it should not throw an exception 😜and successfully add the item to the store.

## Minimal reproduction of the problem with instructions
Simply try adding a document with a field that is a firestore.DocumentReference.
https://akita-firebase-issue.stackblitz.io

## Environment

<pre><code>
Angular version: 8.0.3 (Seems to still be an issue in 8.2.14

Browser:
- [x] Chrome (desktop) version 79.0.3945.88
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
</code></pre>

**Comment by dylanwatsonsoftware** [comment]
https://github.com/dappsnation/akita-ng-fire/issues/43 seems to be a similar issue but this is definitely not an issue with akita-ng-fire.

**Comment by NetanelBasal** [comment]
It seems related to #348. I don't work with firebase, but it seems that you're trying to save the `DocumentReference` object, which feels wrong. IMHO, You need to save the object value.

@GrandSchtroumpf any thoughts on this?

**Comment by GrandSchtroumpf** [comment]
My guess is that `DocumentReference` is a [class instance](https://firebase.google.com/docs/reference/js/firebase.firestore.DocumentReference) with deep properties (reference to the firestore singleton...) and not a plain object.
You can use the [Store Middlewares](https://netbasal.gitbook.io/akita/general/store-middleware) to work around it : 
```typescript
akitaPreAddEntity(book: Book) {
  book.ref ? { ...book, ref: book.ref.path } : book; 
}
```

**Comment by NetanelBasal** [comment]
There is no need to save it in the store. @GrandSchtroumpf there is a method that returns the entity's raw value?

**Comment by GrandSchtroumpf** [comment]
Sorry I read to quickly the issue. 
Yes, you should never put a DocumentReference inside the store, but the latest data.
With observable: 
```typescript
this.fire.doc('...').valueChanges().subscribe(v => this.store.upsert(v))
```
With snapshot: 
```typescript
this.fire.doc('...').ref.get().then(snapshot => this.store.upsert(snapshot.id, snapshot.data()))
```

**Comment by NetanelBasal** [comment]
@dylanwatsonsoftware ^^

**Comment by dylanwatsonsoftware** [comment]
Thanks for the quick reply!

Hmm seems like a reasonable though except that means I lose out on 2 things:
* Lazy evaluation of linked entities that I may or may not ever care about (The reason I am using a DocumentReference)
* The model stored in firebase will not match the model used in Angular anymore (Unless I migrate all my data to store paths instead of actual References)

My use case is something like:
```
interface Role {
  company: DocumentReference;
}
```
which I have a firebase collection (and associated store) for and then that DcumentReference links to a:
```
interface Company {
  // ...company fields
}
```

I guess I can change the type to always be a string on the frontend using `akitaPreAddEntity` and ensure that I convert all the DocumentReferences to paths or ids before putting them in the store but that doesn't really help with creating documents that have DocumentReferences.

**Comment by NetanelBasal** [comment]
I'm not familiar with firebase, but I think it's not related directly to Akita. Even with ngrx, you'll not save a `DocumentReference` in the store. 

@GrandSchtroumpf is a firebase expert. Maybe he could give you some tips on how to handle this.

**Comment by dylanwatsonsoftware** [comment]
Ok I managed to solve my problem by implementing/overriding a combination of the `akitaPreAddEntity` hook on my store and the akita-ng-fire `preFormat` hook on my service. Once I've proven out Akita on my existing project, I'll probably look at migrating the data model to not have nested `DocumentReference`s.. but it gets me moving without a large scale data migration just to migrate to using Akita.

For others who come across this, my store and service looks a bit like:

```
@Injectable({ providedIn: "root" })
@CollectionConfig({ path: "roles" })
export class RoleService extends CollectionService<RoleState> {
  constructor(store: RoleStore) {
    super(store);
  }

  protected preFormat<E extends Role>(document: Readonly<Partial<E>>): E {
    return {
      ...document,
      company: <any>this.db.doc(document.company)
    };
  }
}
```
and
```
@Injectable({ providedIn: "root" })
@StoreConfig({ name: "role" })
export class RoleStore extends EntityStore<RoleState> {
  constructor() {
    super();
  }

  akitaPreAddEntity(entity: Role) {
    return {
      ...entity,
      company: typeof entity.company === "string" ? entity.company : (<DocumentReference>entity.company).id
    };
  }
```

So basically for each field in my firebase model that is a `DocumentReference` I convert it to a string on the way into the store (EntityStore#akitaPreAddEntity) and convert it to a `DocumentReference` on the way into firebase (CollectionService#preFormat).

The better solution will down the line will be to convert the existing data.

**Comment by NetanelBasal** [comment]
@dylanwatsonsoftware, this is great, thanks. You're welcome to join Akita's Gitter channel for any further questions you may have.
